### PR TITLE
Added granular testing, renamed for better visibility and set a threshold for regular swaps

### PR DESCRIPTION
**Pull request by NeoZ666**

### CODE REVIEW

File: src/wallet/api.rs [hunk 1]
```diff
@@ -1332,7 +1332,9 @@ impl Wallet {
             change_cost: cost_of_change,
             avg_input_weight,
             avg_output_weight,
-            min_change_value: 100,
+            // The purpose of keeping it hardcoded is 2 fold. It gives a minimum threshold for Privacy Regular Swaps and solves for the check in SpendAPI against minimal dust and subtracting by fee_wchange - since our current coinselection doesn't handle multiple change outputs.
+            // Instead of hardcoding it, we can definitely do it in a smarter manner.
```
Comment (mojoX911) [inline src/wallet/api.rs]: Shorten this comment and explain why 294 * 5 is selected as. Why not any other values?

What is "Privacy Regular Swaps"? Give it a better name. Simplify the comment to explain the effect of this minimal change amount in coinselection.

Add a TODO mentioning it to be improved later.

File: tests/funding_dynamic_splits.rs [hunk 2]
```diff
@@ -52,22 +55,49 @@ fn test_create_funding_txn_with_varied_distributions() {
...
-    taker.get_wallet_mut().sync_no_fail();
-
     // Generate 5 random addresses from the taker's wallet
     let mut destinations: Vec<Address> = Vec::with_capacity(5);
     for _ in 0..5 {
         let addr = taker.get_wallet_mut().get_next_external_address().unwrap();
         destinations.push(addr);
     }
 
+    test_assert_outputs_inputs(taker, destinations.clone(), fee_rate, TARGETS[0]);
```
Comment (mojoX911) [inline tests/funding_dynamic_splits.rs]: This is only testing the first case? What about the other cases?

File: tests/funding_dynamic_splits.rs [hunk 3]
```diff
@@ -101,23 +131,190 @@ fn test_create_funding_txn_with_varied_distributions() {
...
 
-                    println!("outputs: {out_values:?}");
+                    println!("Outputs: {outputs:?}");
                 }
             }
             Err(e) => {
                 println!("Failed to create funding transactions: {e:?}");
             }
         }
     }
```
Comment (mojoX911) [inline tests/funding_dynamic_splits.rs]: Just unwrap. We don't handle errors in tests.

File: tests/funding_dynamic_splits.rs [hunk 4]
```diff
@@ -101,23 +131,190 @@ fn test_create_funding_txn_with_varied_distributions() {
...
+                balance.spendable.to_sat()
+            );
+            let expected_total = UTXO_SETS[0].iter().sum::<u64>();
+            assert_eq!(
+                balance.spendable.to_sat(),
+                expected_total,
+                "Total wallet balance does not match expected total from UTXO set."
+            );
+        }
+        Err(e) => println!("Failed to get wallet balance: {e:?}"),
```
Comment (mojoX911) [inline tests/funding_dynamic_splits.rs]: same. unwrap.

File: tests/funding_dynamic_splits.rs [hunk 5]
```diff
@@ -101,23 +131,190 @@ fn test_create_funding_txn_with_varied_distributions() {
...
+                    "Fee rate ({}) is less than 98% of MIN_FEE_RATE ({})",
+                    actual_feerate,
+                    MIN_FEE_RATE
+                );
+            }
+        }
+        Err(e) => {
+            println!("Failed to create funding transactions: {e:?}");
+        }
+    }
```
Comment (mojoX911) [inline tests/funding_dynamic_splits.rs]: unwrap.

**Review by mojoX911 — CHANGES_REQUESTED** [review]
Need more changes.

File: tests/funding_dynamic_splits.rs [hunk 6]
```diff
@@ -101,23 +131,190 @@ fn test_create_funding_txn_with_varied_distributions() {
...
+                    println!("Inputs : {selected_inputs:?}");
+                    println!("Sum of Inputs: {:?}", {
+                        selected_inputs.iter().map(|a| a.to_sat()).sum::<u64>()
+                    });
 
-                    let out_values = tx.output.iter().map(|o| o.value).collect::<Vec<_>>();
+                    let outputs = tx.output.iter().map(|o| o.value).collect::<Vec<_>>();
 
-                    println!("outputs: {out_values:?}");
+                    println!("Outputs: {outputs:?}");
```
Comment (mojoX911) [inline tests/funding_dynamic_splits.rs]: Just add the asserts for these vectors here, given the target index, different target case should have different vectors. No need for another large testing function.

File: src/wallet/api.rs (hunk 1 above)
Comment (NeoZ666) [reply src/wallet/api.rs]: Done

File: tests/funding_dynamic_splits.rs (hunk 2 above)
Comment (NeoZ666) [reply tests/funding_dynamic_splits.rs]: Below

File: tests/funding_dynamic_splits.rs (hunk 3 above)
Comment (NeoZ666) [reply tests/funding_dynamic_splits.rs]: Done

File: tests/funding_dynamic_splits.rs (hunk 5 above)
Comment (NeoZ666) [reply tests/funding_dynamic_splits.rs]: DOne

File: tests/funding_dynamic_splits.rs (hunk 6 above)
Comment (NeoZ666) [reply tests/funding_dynamic_splits.rs]: I made the larger function to assert only particular inputs being chosen everytime with number of targets/changes consistent everytime. I took one UTXO set and a target for granular testing.
The smaller `test_create_funding_txn_with_varied_distributions` is for specific edge cases.

File: tests/funding_dynamic_splits.rs [hunk 7]
```diff
@@ -1,123 +1,207 @@
...
+        // Assert the number of Outputs matches the expected number.
+        assert_eq!(
+            outputs.len(),
+            NUM_OUTPUT_CHUNKS[i] as usize,
+            "Expected {} outputs, got {}",
+            NUM_OUTPUT_CHUNKS[i],
+            outputs.len()
+        );
+
+        // Assert Fee is more than 98% of the expected Fee Rate.
```
Comment (mojoX911) [inline tests/funding_dynamic_splits.rs]:
```suggestion
        // Assert that the actual fee rate is within 2% of MIN_FEE_RATE
```

File: tests/funding_dynamic_splits.rs [hunk 8]
```diff
@@ -1,123 +1,207 @@
...
+        assert!(
+            actual_feerate >= MIN_FEE_RATE * 0.98,
+            "Fee rate ({}) is less than 98% of MIN_FEE_RATE ({})",
+            actual_feerate,
+            MIN_FEE_RATE
+        );
```
Comment (mojoX911) [inline tests/funding_dynamic_splits.rs]: This assert is not very effective, as if the fee rate is very high, it will pass through the clause.

You need a 2% lower and upper bound check.

File: tests/funding_dynamic_splits.rs [hunk 9]
```diff
@@ -1,123 +1,207 @@
...
+        println!("Actual fee rate: {actual_feerate}");
+
+        // Assert no duplicate inputs.
+        assert!(selected_inputs.iter().all(|&x| selected_inputs
+            .iter()
+            .filter(|&&y| y == x)
+            .count()
+            == 1),);
+
+        // Assert the Output UTXOs matches the expected outputs.
```
Comment (mojoX911) [inline tests/funding_dynamic_splits.rs:130]:
```suggestion
        // Assert that the selected UTXOs, matches with the expected INPUT list. 
```

File: tests/funding_dynamic_splits.rs [hunk 10]
```diff
@@ -1,123 +1,207 @@
...
+        // Call `create_funding_txes` with Normie Flag turned off, i.e Destination::MultiDynamic
+        let result = taker
+            .get_wallet_mut()
+            .create_funding_txes_regular_swaps(
+                false,
+                target,
+                destinations.clone(),
+                Amount::from_sat(MIN_FEE_RATE as u64),
+            )
+            .unwrap();
```
Comment (mojoX911) [inline tests/funding_dynamic_splits.rs:89]: Eventually the normie flag will be redundant, as we can dictate the type of the spend with `DestinationType`. That can be done in a followup PR.

Will make the spend APIs simpler.

Instead of `MultiDynamic`,  lets name it something more meaningful. Like `UtxoSplitter` or similar. With a good doc explaining the behavior in the Destination enum variant site.

This can be done in a followup PR.

**Review by mojoX911 — CHANGES_REQUESTED** [review]
Going towards the right direction. Test is failing on utxo mismatch. Some nits and followup PR ideas.

File: tests/funding_dynamic_splits.rs (hunk 8 above)
Comment (NeoZ666) [reply tests/funding_dynamic_splits.rs]: It overshoots in one of the case to 2.41

File: tests/funding_dynamic_splits.rs (hunk 10 above)
Comment (NeoZ666) [reply tests/funding_dynamic_splits.rs:89]: Makes sense. We can cover these fixes + a doc for explaining the entire behavior and flow.

**Comment by NeoZ666** [comment]
The test will fail owing to pending to-be-merged changes in CS (main<-temp) @mojoX911

**Comment by mojoX911** [comment]
Revert the commits in `cs::main`, the amounts will fail as expected. Rebase this PR to `tmp` branch first, make all tests green, then we can merge to `cs::main`.

**Comment by NeoZ666** [comment]
Rebased on top of #558

**Review by mojoX911 — APPROVED** [review]
Ack.

**Review by Wukong247 — APPROVED** [review]
